[DONE] Refactoring the photoprism package.

Skarlso opened the pull request:
This relates to #50.

Works still to be done:
- [x] Some files are left out yet and not finished I'm just sharing work
- [x] Revise the `util` file and throw it out
- [x] See other exported functions if they need to be exported
~- [ ] Revise the use of pointer structs instead of value. If they aren't changed there is no need for pointers to needlessly occupy heap.~ -- this would actually be a ton of work right now. I decided to leave that off to a different PR.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/photoprism/photoprism?pullRequest=53) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/photoprism/photoprism?pullRequest=53) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/photoprism/photoprism?pullRequest=53) it.</sub>

lastzero commented on internal/photoprism/colors.go:
```diff
@@ -1,12 +1,13 @@
 	"image"
 	"os"
 	"sort"
+
```
Please make sure to run `make fmt` to avoid changes in the import order (should sort them alphabetically)

Skarlso replied:
These import statements have been formatted by the Go standard, `goformat`. Fmt states this about it:

>gofmt follows a simple rule that works for all use cases: imports may be separated by
blank lines, and each set of blank lines is sorted independently.
goimports, which does more than just formatting, applies a different rule: it groups
imports according to whether they are standard imports, appengine imports, or other
imports.  It emits each group separately.  Then the gofmt algorithm kicks in and sorts
each group independently.
The goimports and gofmt programs do different jobs and it is not a goal that they
produce identical results.

Are you saying you don't want `goformat`? :)

Also, instead of a target your IDE should definitely take care of formatting rules as that's a much more convenient thing to have instead of running a make target.

lastzero replied:
You mean `goimports`, not `goformat`? `make fmt` uses `go fmt`.

If `goimports` doesn't cause any issues (does it need installation? will GoLand remove the empty line again?), we can add it to the make target. Just want the imports to be the same for everyone.

lastzero replied:
If everyone uses different IDE settings, merge conflicts and differently formatted code will be the result. That's why we use make. Of course if your IDE can do it the same, use it. Maybe I should also check my IDE settings... not a bad idea.

lastzero replied:
Added it to `make fmt`, see 6a2af16d1cd556d4d25eb2c43c1cd43e9b9dad71 / #50

Skarlso replied:
Interesting. I thought goland uses these tools by default. By the way that's why I like Go, formatting is a language standard and thus every tool from vimgo to goland should be applying them.

Yes, you are right. 😊 I meant goimports.

lastzero replied:
No, you need to manually install `goimports` and enable it as a save hook. Should add that to the developer guide. Not a huge fan of extra tools in general: https://twitter.com/lastzero/status/1056992297135603712

lastzero commented on internal/photoprism/config.go:
```diff
@@ -1,38 +1,45 @@
+	ServerPort     int    `yaml:"server-port"`
+	ServerMode     string `yaml:"server-mode"`
+	AssetsPath     string `yaml:"assets-path"`
+	ThumbnailsPath string `yaml:"thumbnails-path"`
```
As discussed in #50, I think it's better to use functions in this particular case. I'll try to document that for others inline and/or in the developer guide. Thank you for the valuable feedback and suggestions!

Skarlso commented:
@lastzero I got the email but I can't see your comment any longer. :O

So, the util is usually anti pattern in a code. It's code no-one know where it should go. But these were usually functions that were used in either one place or actually belonged together like a file_checker. Which were two common functions that did file operations.

The rest was only used in one place and those belongs to that place. Also, some of these functions wasn't used at all, or only in the test which made no sense to me. It was dead code. So I deleted it. Like the random generator.

If you require it it's always under version control. :) That's why I removed it. :)

lastzero commented:
In final code it would be like that, I agree. The util file was a reminder to use a proper abstraction like spf13/afero later (and even that puts them in util.go...). Like many rough edges are actually reminders for something bigger, because they jump in your eye (like they did! see it worked...). Now the functions are distributed in different files :) But it's ok.

lastzero commented:
My focus will be to improve documentation and tests before I start adding new features. Very important when you work in a distributed team. When there are tests, you can always refactor later.

lastzero commented on internal/photoprism/search.go:
```diff
@@ -177,26 +182,31 @@ func (s *Search) Photos(form PhotoSearchForm) ([]PhotoSearchResult, error) {
-func (s *Search) FindPhotoById(photoId uint64) (photo Photo) {
-	s.db.Where("id = ?", photoId).First(&photo)
+// FindPhotoByID returns a Photo based on the ID.
+func (s *Search) FindPhotoByID(photoID uint64) (photo models.Photo) {
```
Ah, didn't work because you changed the imports ;)

Skarlso commented:
@lastzero Cool stuff man! I take your view into account on the next changes! :) 👍
